File: test/resources/repos/fsharp/test_repo/src/Calculator.fs
module TestProject.Calculator

open TestProject.Helper
open TestProject.Types

/// A calculator class that performs basic arithmetic operations
type Calculator() =

    /// Add two numbers
    member this.Add(x, y) = add x y

    /// Subtract two numbers
    member this.Subtract(x, y) = subtract x y

    /// Multiply two numbers
    member this.Multiply(x, y) = multiply x y

    /// Divide two numbers
    member this.Divide(x, y) = divide x y

    /// Calculate the sum of a list of numbers
    member this.Sum(numbers: int list) =
        List.fold (fun acc n -> add acc n) 0 numbers

    /// Calculate the product of a list of numbers
    member this.Product(numbers: int list) =
        List.fold (fun acc n -> multiply acc n) 1 numbers

/// Create a calculator instance
let createCalculator() = Calculator()

/// Evaluate an expression using the calculator
let evaluate (calc: Calculator) op x y =
    match op with
    | "add" -> calc.Add(x, y)
    | "subtract" -> calc.Subtract(x, y)
    | "multiply" -> calc.Multiply(x, y)
    | "divide" -> calc.Divide(x, y)
    | _ -> failwith "Unknown operation"

File: test/resources/repos/nim/test_repo/src/calculator.nim
## Calculator module with basic and advanced arithmetic operations.

proc add*(a, b: float): float =
  ## Add two numbers.
  result = a + b

proc subtract*(a, b: float): float =
  ## Subtract b from a.
  result = a - b

proc multiply*(a, b: float): float =
  ## Multiply two numbers.
  result = a * b

proc divide*(a, b: float): float =
  ## Divide a by b. Raises DivByZeroDefect if b is zero.
  if b == 0.0:
    raise newException(DivByZeroDefect, "Cannot divide by zero")
  result = a / b

proc factorial*(n: int): int =
  ## Compute factorial of a non-negative integer.
  if n < 0:
    raise newException(ValueError, "Factorial not defined for negative numbers")
  if n <= 1:
    return 1
  result = n * factorial(n - 1)

proc mean*(values: seq[float]): float =
  ## Compute the arithmetic mean of a sequence of floats.
  if values.len == 0:
    raise newException(ValueError, "Cannot compute mean of empty sequence")
  var total = 0.0
  for v in values:
    total += v
  result = total / float(values.len)
